[compiler][rewrite] PropagateScopeDeps hir rewrite

mofeiZ opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #30894
* #30949
* #30948

Resubmission of #30079 -- core logic unchanged, but needed to rebase past #30573

### Quick background
#### Temporaries

The compiler currently treats temporaries and named variables (e.g. `x`) differently in this pass.
- named variables may be reassigned (in fact, since we're running after LeaveSSA, a single named identifier's IdentifierId may map to multiple `Identifier` instances -- each with its own scope and mutable range)
- temporaries are replaced with their represented expressions during codegen. This is correct (mostly correct, see #29878) as we're careful to always lower the correct evaluation semantics. However, since we rewrite reactive scopes entirely (to if/else blocks), we need to track temporaries that a scope produces in `ReactiveScope.declarations` and later promote them to named variables.
In the same example, $4, $5, and $6 need to be promoted: $2 ->`t0`,  $5 ->`t1`, and $6 ->`t2`.
```js
[1] $2 = LoadGlobal(global) foo
[2] $3 = LoadLocal bar$1
scope 0:
  [3] $4 = Call $2(<unknown> $3)
scope 1:
  [4] $5 = Object {  }
scope 2:
  [5] $6 = Object { a: $4, b: $5 }
[6] $8 = StoreLocal Const x$7 = $6
```

#### Dependencies
`ReactiveScope.dependencies` records the set of (read-only) values that a reactive scope is dependent on. This is currently limited to just variables (named variables from source and promoted temporaries) and property-loads.
All dependencies we record need to be hoistable -- i.e. reordered to just before the ReactiveScope begins. Not all PropertyLoads are hoistable.

In this example, we should not evaluate `obj.a.b` without before creating x and checking `objIsNull`.
```js
// reduce-reactive-deps/no-uncond.js
function useFoo({ obj, objIsNull }) {
  const x = [];
  if (isFalse(objIsNull)) {
    x.push(obj.a.b);
  }
  return x;
}
```

While other memoization strategies with different constraints exist, the current compiler requires that `ReactiveScope.dependencies` be re-orderable to the beginning of the reactive scope. But.. `PropertyLoad`s from null values will throw `TypeError`. This means that evaluating hoisted dependencies should throw if and only if the source program throws. (It is also a bug if source throws and compiler output does not throw. See https://github.com/facebook/react-forget/pull/2709)

---
### Rough high level overview
1. Pass 1
Walk over instructions to gather every temporary used outside of its defining scope (same as ReactiveFunction version). These determine the sidemaps we produce, as temporaries used outside of their declaring scopes get promoted to named variables later (and are not considered hoistable rvals).
2. Pass 2 (collectTemporariesSidemap)
Walk over instructions to generate a sidemap of temporary identifier -> named variable and property path (e.g. `$3 -> {obj: props, path: ["a", "b"]}`)
2. Pass 2 (collectHoistablePropertyLoads)
  a. Build a sidemap of block -> accessed variables and properties (e.g. `bb0 -> [ {obj: props, path: ["a", "b"]} ]`)
  b. Propagate "non-nullness" i.e. variables and properties for which we can safely evaluate `PropertyLoad`.
  A basic block can unconditionally read from identifier X if any of the following applies:
    - the block itself reads from identifier X
    - all predecessors of the block read from identifier X
    - all successors of the block read from identifier X
4. Pass 3: (collectDependencies)
Walks over instructions again to record dependencies and declarations, using the previously produced sidemaps. We do not record any control-flow here
5. Merge every scope's recorded dependencies with the set of hoistable PropertyLoads

Tested by syncing internally and (1) checking compilation output differences ([internal link](https://www.internalfb.com/intern/everpaste/?handle=GPCfUBt_HCoy_S4EAJDVFJyJJMR0bsIXAAAB)), running internally e2e tests ([internal link](https://fburl.com/sandcastle/cs5mlkxq))

---
### Followups:
1. Rewrite function expression deps
This change produces much more optimal output as the compiler now uses the function CFG to understand which variables / paths are assumed to be non-null. However, it may exacerbate [this function-expr hoisting bug](https://github.com/facebook/react/blob/main/compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-invalid-hoisting-functionexpr.tsx). A short term fix here is to simply call some form of `collectNonNullObjects` on every function expression to find hoistable variable / paths. In the longer term, we should refactor out `FunctionExpression.deps`.

2. Enable optional paths
(a) don't count optional load temporaries as dependencies (e.g. `collectOptionalLoadRValues(...)`).
(b) record optional paths in both collectHoistablePropertyLoads and dependency collection

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,517 @@
+  assumedNonNullObjects: Set<PropertyLoadNode>;
+};
+
+export function getProperty(
```
`getProperty` and `resolveTemporary` were essentially moved from PropagateScopeDeps (reactive scopes version). Understanding hoistable property loads required the same `properties` and `temporaries` sidemaps as the core dependency-extraction logic. Now, we first compute the sidemaps, then use them (as `ReadOnlyMaps`) in both `CollectHoistablePropertyLoads` and `PropagateScopeDeps`

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,517 @@
+  propertyLoadInfo: Map<BlockId, Array<PropertyLoadInfo>>;
+};
+
+function collectSidemap(
```
(copied logic from PropagateScopeDeps)

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,517 @@
+  fn: HIRFunction,
+  sidemap: TemporariesSidemap,
+): Map<BlockId, BlockInfo> {
+  /**
+   * Due to current limitations of mutable range inference, there are edge cases in
+   * which we infer known-immutable values (e.g. props or hook params) to have a
+   * mutable range and scope.
+   * (see `destructure-array-declaration-to-context-var` fixture)
+   * We track known immutable identifiers to reduce regressions (as PropagateScopeDeps
+   * is being rewritten to HIR).
+   */
+  const knownImmutableIdentifiers = new Set<Identifier>();
```
This patches the the regression from fixtures like `destructure-array-declaration-to-context-var` (see [comment on original PR](https://github.com/facebook/react/pull/30079#discussion_r1675390138))

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,517 @@
+  fn: HIRFunction,
+  sidemap: TemporariesSidemap,
+): Map<BlockId, BlockInfo> {
+  /**
+   * Due to current limitations of mutable range inference, there are edge cases in
+   * which we infer known-immutable values (e.g. props or hook params) to have a
+   * mutable range and scope.
+   * (see `destructure-array-declaration-to-context-var` fixture)
+   * We track known immutable identifiers to reduce regressions (as PropagateScopeDeps
+   * is being rewritten to HIR).
+   */
+  const knownImmutableIdentifiers = new Set<Identifier>();
+  if (fn.fnType === 'Component' || fn.fnType === 'Hook') {
+    for (const p of fn.params) {
+      if (p.kind === 'Identifier') {
+        knownImmutableIdentifiers.add(p.identifier);
+      }
+    }
+  }
```
This patches the the regression from fixtures like `destructure-array-declaration-to-context-var` (see [comment on original PR](https://github.com/facebook/react/pull/30079#discussion_r1675390138))

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,517 @@
+       *
+       * See comment at top of function for why we track known immutable identifiers.
+       */
+      const isMutableAtInstr =
+        object.mutableRange.end > object.mutableRange.start + 1 &&
+        object.scope != null &&
+        inRange(instr, object.scope.range);
```
This is a little awkward, and also an example of being able to make use of more granular mutable range (instead of scope-range info)

Note that mutableRanges are no longer valid at this point, but we're only comparing relative start/ends (end === start + 1 indicates nonmutable value)

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/DeriveMinimalDependenciesHIR.ts:
```diff
@@ -0,0 +1,267 @@
+  }
+}
+
+enum PropertyAccessType {
+  Access = 'Access',
+  NonNullAccess = 'NonNullAccess',
+  Dependency = 'Dependency',
+  NonNullDependency = 'NonNullDependency',
+}
+
+const MIN_ACCESS_TYPE = PropertyAccessType.Access;
+/**
+ * "NonNull" means that PropertyReads from a node are side-effect free,
+ * as the node is (1) immutable and (2) has unconditional propertyloads
+ * somewhere in the cfg.
+ */
+function isNonNull(access: PropertyAccessType): boolean {
+  return (
+    access === PropertyAccessType.NonNullAccess ||
+    access === PropertyAccessType.NonNullDependency
+  );
+}
+function isDependency(access: PropertyAccessType): boolean {
+  return (
+    access === PropertyAccessType.Dependency ||
+    access === PropertyAccessType.NonNullDependency
+  );
+}
```
Note this is slightly different from the original DeriveMinimalDeps.

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/DeriveMinimalDependenciesHIR.ts:
```diff
@@ -0,0 +1,267 @@
+  }
+}
+
+function deriveMinimalDependenciesInSubtree(
```
Note that this logic is now much simpler than source [`DeriveMinimalDependencies`](https://github.com/facebook/react/blob/main/compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/DeriveMinimalDependencies.ts#L335). The key change is that we know the tree is a 'well formed tree' -- if any child nodes of a node are non-null (or unconditional), that node itself must be non-null.

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:
```diff
@@ -0,0 +1,512 @@
+  }
+}
+
+function findTemporariesUsedOutsideDeclaringScope(
```
Copied from PropagateScopeDeps

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:
```diff
@@ -0,0 +1,512 @@
+  #declarations: Map<DeclarationId, Decl> = new Map();
+  #reassignments: Map<Identifier, Decl> = new Map();
+  // Reactive dependencies used in the current reactive scope.
+  #dependencies: Stack<Array<ReactiveScopeDependency>> = empty();
```
Note that `dependencies is now a stack since we no longer are recursing into scopes. We only ever use the top entry of this, so I'm happy to add/use a new array based utility class that only exposes push, pop, and top

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:
```diff
@@ -0,0 +1,512 @@
+    this.#properties = properties;
+  }
+
+  enterScope(scope: ReactiveScope): void {
+    // Set context for new scope
+    this.#dependencies = this.#dependencies.push([]);
+    this.#scopes = this.#scopes.push(scope);
+  }
+
+  exitScope(scope: ReactiveScope, pruned: boolean): void {
```
Same logic as reactive scope based PropagateScopeDeps

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:
```diff
@@ -0,0 +1,512 @@
+      }
+    }
+
+    if (!pruned) {
+      this.deps.set(scope, scopedDependencies);
+    }
```
Note that here, we store a scope's dependencies to later merge with results from `collectHoistablePropertyLoads`. The ordering doesn't really matter here, as both scope dependencies and non-null objects are read-only.

If this is confusing, I can move the merging logic here

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:
```diff
@@ -0,0 +1,512 @@
+   * scope we are visiting is the same scope where the value originates, it can't be a dependency
+   * on itself.
+   */
+  declare(identifier: Identifier, decl: Decl): void {
+    if (!this.#declarations.has(identifier.declarationId)) {
+      this.#declarations.set(identifier.declarationId, decl);
+    }
+    this.#reassignments.set(identifier, decl);
+  }
+
+  resolveTemporary(place: Place): Identifier {
+    return this.#temporaries.get(place.identifier) ?? place.identifier;
+  }
+
+  // Checks if identifier is a valid dependency in the current scope
+  #checkValidDependency(maybeDependency: ReactiveScopeDependency): boolean {
```
Copied from propagateScopeDeps (reactive scopes version)

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:
```diff
@@ -0,0 +1,512 @@
+   * scope we are visiting is the same scope where the value originates, it can't be a dependency
+   * on itself.
+   */
+  declare(identifier: Identifier, decl: Decl): void {
```
All below methods are copied from propagateScopeDeps (reactive scopes version). The only change is that I cleaned up all `isConditional` / non-ast control flow (return / throw) logic, and moved temporaries / properties sidemap creation to an earlier step

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:
```diff
@@ -0,0 +1,512 @@
+  }
+}
+
+function handleInstruction(instr: Instruction, context: Context): void {
```
(copied from propagateScopeDeps)

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/visitors.ts:
```diff
@@ -1149,3 +1153,72 @@ export function* eachTerminalOperand(terminal: Terminal): Iterable<Place> {
+  }
+
+  isScopeActive(scopeId: ScopeId): boolean {
+    return this.#activeScopes.indexOf(scopeId) !== -1;
```
I believe the `indexOf` check should be cheap, as # of nested scopes is generally low. Happy to add a backing set if this isn't true though

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,517 @@
+  const nodes = collectNodes(fn, sidemap);
+  deriveNonNull(fn, nodes);
+
+  const nodesKeyedByScopeId = new Map<ScopeId, BlockInfo>();
+  for (const [_, block] of fn.body.blocks) {
+    if (block.terminal.kind === 'scope') {
+      nodesKeyedByScopeId.set(
+        block.terminal.scope.id,
+        nodes.get(block.terminal.block)!,
+      );
+    }
+  }
```
> maybe we should translate nodes into a Map keyed by scope earlier

(from @josephsavona -- done!)

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,517 @@
+  return nodes;
+}
+
+function deriveNonNull(fn: HIRFunction, nodes: Map<BlockId, BlockInfo>): void {
```
> Gotcha. That makes sense. I suspect this can still be simplified - in particular, i had to remap the variable names in my head to help keep track of things. See comments for the most unclear cases.
> 
> One thing that strikes me about this algorithm is that it isn't accurate in the sense that it doesn't account for the actual live range of values. During the initial collectNodes stage we account for mutable ranges when deciding if a property access can locally be considered to be non-nullable. But then in the deriveNonNull phase we don't have access to the mutable ranges anymore. AFAICT, if you have:
> 
```
bb0:
  let x = null
  if cond then bb1 else bb2
bb1:
  x = {}
  goto bb3
bb2:
  ...
  goto bb3:
bb3:
  x.foo.bar
```
> 
> this algorithm will infer `x.foo.bar` as non-nullable for bb0, via the intersection of bb1 & bb2 which in turn comes from bb3. So the results are actually inaccurate, but that's okay because of how we filter out dependencies in PropagateScopeDeps. 
> 
> This isn't necessarily a blocker, but it isn't great to have a pass that returns results that are technically incorrect — it's very easy to accidentally use them in a different way later, w/o realizing that important context. Thoughts on this, and whether we could make the results guaranteed accurate?

([comment from @josephsavona](https://github.com/facebook/react/pull/30079#discussion_r1675380685))

I added a comment at the top of file (on the exported function) to note this technical inaccuracy. I'm inclined to leave it for now, as reusing this logic comes with other potential gotchas (e.g. use of an identifier outside of block scope, use after reassignment, etc)

mofeiZ replied:
(This follows a non-dominator based approach)

Take this example. Here, we know that it's safe to access `a.b` on line 5 even though no dominator of its basic block accesses a.b.
```js
0: if(...) {
1:   a.b;
2: } else {
3:   a.b;
4: }
5: // It's safe to take `a.b` as a dependency of scope@0 (i.e. hoist the a.b access)
6: scope @0 {
7:   if (cond) {
8:     // access a.b;
9:   }
10:}
```

Currently, we record that `a.b` is safe to read in $bb_i$ if any of the following hold:
- $bb_i$ itself accesses a.b
- a.b is safe to access in every pred of `a.b`
- a.b is safe to access in every successor of `a.b`

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,517 @@
+      direction === 'backward' ? (succ.get(nodeId) ?? []) : node.block.preds,
+    );
+
+    let changed = false;
+    for (const pred of neighbors) {
+      if (!traversalState.has(pred)) {
+        const neighborChanged = recursivelyDeriveNonNull(
+          pred,
+          direction,
+          traversalState,
+          nonNullObjectsByBlock,
+        );
+        changed ||= neighborChanged;
+      }
+    }
+    /**
+     * Active neighbors can be filtered out as we're solving for the following
+     * relation.
+     * X = Intersect(X_neighbors, X)
+     * Non-active neighbors with no recorded results can occur due to backedges.
+     * it's not safe to assume they can be filtered out (e.g. not intersected)
+     */
+    const neighborAccesses = Set_intersect([
+      ...(Array.from(neighbors)
+        .filter(n => traversalState.get(n) === 'done')
+        .map(n => nonNullObjectsByBlock.get(n) ?? new Set()) as Array<
+        Set<PropertyLoadNode>
+      >),
+    ]);
```
This is the core logic of the `deriveNonNull`. We first derive non-null objects in neighbors (depending on direction), then calculate `X = Intersect(...X_neighbors, X)`

mofeiZ replied:
[Tree.getPropertyLoadNode](https://github.com/facebook/react/blob/999bb6bebcd83570c936fe6f25f0b15de5000140/compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts#L221-L238) uniques all property loads

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,517 @@
+    );
+    traversalState.set(nodeId, 'done');
+
+    changed ||= prevSize !== nonNullObjectsByBlock.get(nodeId)!.size;
+    CompilerError.invariant(
+      prevSize == null || prevSize <= nonNullObjectsByBlock.get(nodeId)!.size,
+      {
+        reason: '[CollectHoistablePropertyLoads] Nodes shrank!',
+        description: `${nodeId} ${direction} ${prevSize} ${
+          nonNullObjectsByBlock.get(nodeId)!.size
+        }`,
+        loc: GeneratedSource,
+      },
+    );
+    return changed;
```
Will remove the fixed-point iteration logic in a later PR after more testing internally

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/reduce-reactive-deps/hoist-deps-diff-ssa-instance1.expect.md:
```diff
@@ -0,0 +1,96 @@
+    x = t2;
+  }
+  let t2;
+  if ($[2] !== cond || $[3] !== x) {
```
This is finally correct -- fixed by #30573 ! Note that before removing `LeaveSSA`, we incorrectly take a dependency on `x` here as we weren't able to distinguish reassignment instances (see [previous repro on previous PR](https://react-compiler-playground-fboeug9zp-fbopensource.vercel.app/#N4Igzg9grgTgxgUxALhAHQHYDMobgFwEsIMACAMQggAphS4SATUgX2VLoY0fYCMqANggCGZFgEoOmUqSH5SAD3YAJYWACCpAD6lVYAMKkAvKTAALYQIEQA7vogAHAJ61SAa0Ld2AcgsbvADSkwux0AG6WUAjsAEysrOIA3NKkAPQAVOkppOmkACpmhGCkMCLMEFiKAHTCVRECUaRFpND4YISMCC2VCt7FHQgYRFiECDCkALZQ+MK8Qtm5MKIA5l0A1KYMDl1LGKtVpADqXRPCMG6k+IXFYGDCTRhgM3hd1AoA+gACAAySaqQWMKeZYLYKkDAkAC0GCgVhavAAVggCKQHDBHGN8E5SAADBQ1HFVBapFIAWWEVyqpwUbxqdUiCCCMSSKUIlWoXEYkmA2UUxlMFistnszlcHi8pF8an0gXooVI9Si7AAzPEJMkyKxMCkMllNTlSABlTyIS7XPlvL5xEwOQqWn5BD6fACM4kkzXujDZWDGg3kt3unieokQAVBNi6cFEEP9wh9AmxZggRXk91KwnKlTxdMVCBxxJSpXwsDIAB5DfgYMC2dj6kZgABCTmkABkLdIAG18bVc6QNjEALosNIAPg1LBALCAA))

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructure-array-assignment-to-context-var.expect.md:
```diff
@@ -30,15 +30,15 @@ import { identity } from "shared-runtime";
   const $ = _c(4);
   let x;
-  if ($[0] !== props.value) {
+  if ($[0] !== props) {
```
Hmmm why does this regress? `props.value` is unconditionally accessed both at the top-level and in the function expression so i'd expect that to be the inferred dep

mofeiZ replied:
This is the same case as https://github.com/facebook/react/pull/30894#discussion_r1747781289, will change to return jsx. Happy to also improve `compilationMode: 'all'` in a followup

josephsavona replied:
got it, sounds good!

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructure-array-declaration-to-context-var.js:
```diff
@@ -6,7 +6,7 @@ function Component(props) {
   };
   foo();
-  return {x};
+  return <div>{x}</div>;
```
why do we need to change here? this would force it to freeze, but why is that relevant?

mofeiZ replied:
Ahh this is kind of an edge case. It's not necessary to freeze x before the return statement. I needed to add either jsx or a hook call to get `snap` to infer that this function is a `Component`|`Hook` (instead of `Other` type). This is not a problem internally as we rely on component syntax, so every function we attempt to compile is guaranteed to be a component or hook. It's also not an issue for `compilationMode: "infer"` cases without component syntax. Snap, however, uses `compilationMode: 'all'` by default. This mode should probably be patched to infer functions that don't create jsx / call hooks by function naming.

We filter by fnKind = 'Hook' | 'Component' when computing known immutable identifiers [here](https://github.com/facebook/react/blob/999bb6bebcd83570c936fe6f25f0b15de5000140/compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts#L298-L313)
Without this, we infer a (long) mutable range for `props` due to it being directly aliased into a context variable, which gets captured / mutated within a lambda.

josephsavona replied:
ahhh got it, thanks for clarifying!

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/optional-call-with-optional-property-load.expect.md:
```diff
@@ -15,9 +15,9 @@ import { c as _c } from "react/compiler-runtime";
   const $ = _c(2);
   let t0;
-  if ($[0] !== props?.items) {
+  if ($[0] !== props) {
```
expected since we don't support optionals yet

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/todo-global-load-cached.tsx:
```diff
@@ -0,0 +1,17 @@
+import {Stringify} from 'shared-runtime';
+import {makeArray} from 'shared-runtime';
+
+/**
+ * Here, we don't need to memoize Stringify as it is a read off of a global.
+ * TODO: in PropagateScopeDeps (hir), we should produce a sidemap of global rvals
+ * and avoid adding them to `temporariesUsedOutsideDefiningScope`.
+ */
```
Hmm, I don't think we're "memoizing" Stringify here. Instead, it's inside the mutable range of `arr` and therefore the temporary is actually used outside its declaring scope. 

This would be fixed by InstructionReordering, since it would move the LoadGlobal later to just before the JSX element, getting it out of the scope.

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/todo-global-property-load-cached.tsx:
```diff
@@ -0,0 +1,20 @@
+import * as SharedRuntime from 'shared-runtime';
+import {makeArray} from 'shared-runtime';
+
+/**
+ * Here, we don't need to memoize SharedRuntime.Stringify as it is a PropertyLoad
+ * off of a global.
+ * TODO: in PropagateScopeDeps (hir), we should produce a sidemap of global rvals
+ * and avoid adding them to `temporariesUsedOutsideDefiningScope`.
+ */
```
same as above

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/try-catch-try-value-modified-in-catch-escaping.expect.md:
```diff
@@ -31,9 +31,9 @@ import { c as _c } from "react/compiler-runtime";
+  const $ = _c(2);
   let x;
-  if ($[0] !== props.y || $[1] !== props.e) {
+  if ($[0] !== props) {
```
regression

mofeiZ replied:
@josephsavona I believe that this is more of a bugfix than a regression. Take this example, previously we infered dependencies on `cond` as well as `a.b` and `a.c`. This is because we didn't correctly handle `MaybeThrow` terminals.
```js
function useFoo(a, cond) {
  let x;
  try {
    const y = [];
    throwIfTrue(cond);
    y.push(a.b);
    throwInput(y);
  } catch (t0) {
    const e = t0;
    e.push(a.c);
    x = e;
  }
  return x;
}
// playground output
function useFoo(a, cond) {
  const $ = _c(4);
  let x;
  if ($[0] !== cond || $[1] !== a.b || $[2] !== a.c) {
    try {
      const y = [];
      throwIfTrue(cond);
      y.push(a.b);
      throwInput(y);
    } catch (t0) {
      const t0$0 = t0;
      const e = t0$0;
      e.push(a.c);
      x = e;
    }
    $[0] = cond;
    $[1] = a.b;
    $[2] = a.c;
    $[3] = x;
  } else {
    x = $[3];
  }
  return x;
}
```

[Here](https://0xeac7-forget.vercel.app/#N4Igzg9grgTgxgUxALhAMygOzgFwJYSYAEUYCAYhBABQCGANEXIQCYCURwAOsUQDYIcRAB4BuHkSaEwQgJ5EAvEQDaAXXG8cMed16ScACxgQA7gEk0AFRhQE1ZpnYbJk2QDoADqQN03AIzZnF0Njc0wvHGpZQIkiAF8mWhw4AyJqHAAGDl0XKUwZIgRFIkygyQRPb2oAJhi9EWKEILjYmEFYYjEeOJA4oA) is an another example that is currently just a bug

I agree that we can technically make `MaybeThrow` handling more precise though, will look into it!

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/try-catch-try-value-modified-in-catch.expect.md:
```diff
@@ -30,9 +30,9 @@ import { c as _c } from "react/compiler-runtime";
+  const $ = _c(2);
   let t0;
-  if ($[0] !== props.y || $[1] !== props.e) {
+  if ($[0] !== props) {
```
regression

mofeiZ replied:
(responded on other thread https://github.com/facebook/react/pull/30894/files/999bb6bebcd83570c936fe6f25f0b15de5000140#r1747787204 )

josephsavona reviewed (commented):
First pass of code review just looking at the output changes. Super exciting! This is clearly a better approach, as we've consistently seeing with moving to HIR.

A few themes:
* Mostly more precise, yay!
* The LoadGlobal and PropertyLoads inside scopes is i believe due to what we discussed over VC, it's just an accident of the ordering. The right fix for this is instruction reordering. I wouldn't worry about these cases for now since reordering will take care of them.
* A few cases regress bc of missing support for optionals. 
* But a few other cases appear to regress unrelated to optionals. We should definitely investigate these.

In terms of steps forward: I think it makes sense (after reviewing the rest of the code) to move forward with landing and then iterating on the follow-ups you discussed (function expressions, optionals). But we shouldn't turn the feature flag on by default until that's all ready. Let's fork the interesting fixtures and flip the flag to off by default?

mvitousek commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,520 @@
+export type CollectHoistablePropertyLoadsResult = {
+  nodes: ReadonlyMap<ScopeId, BlockInfo>;
+  temporaries: ReadonlyMap<Identifier, Identifier>;
+  properties: ReadonlyMap<Identifier, ReactiveScopeDependency>;
```
For these sidemaps, any reason to key it by the identifier itself and not an identifierID?

mofeiZ replied:
Updated to use `IdentifierId` and double checked that this changed nothing internally on 100k+ files

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/try-catch-mutate-outer-value.expect.md:
```diff
@@ -28,9 +28,9 @@ import { c as _c } from "react/compiler-runtime";
+  const $ = _c(5);
   let x;
-  if ($[0] !== props.a) {
+  if ($[0] !== props) {
```
Note that this is also a bugfix. I can change the fixture slightly and produce a difference in the evaluator

```js
function useHook(maybeNull) {
  const x = [];
  try {
    x.push(2);
  } catch {
    x.push(shallowCopy(maybeNull.value));
  }
  return x;
}
```

mvitousek commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,520 @@
+  reversedBlocks.reverse();
+  let i = 0;
+
+  while (changed) {
```
If we decide to (or need to, for phis) keep this fixpoint logic, we should break out of it with an error if it goes on for "too long" so we don't diverge if there's a bug introduced someday

mvitousek replied:
Also maybe a do-while so we don't have the bogus 'let changed = true' above

mvitousek commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,520 @@
+    /**
+     * Active neighbors can be filtered out as we're solving for the following
+     * relation.
+     * X = Intersect(X_neighbors, X)
```
This doesn't look like it describes the algorithm. It looks like what we're actually getting is Union(X, Intersect(...X_neighbors)), no? (cf: Set_intersect on line 422, and Set_union on line 433)

mofeiZ replied:
Yep that's totally right, I separated this comment into two -- one describing the general algorithm (the Union(X, Intersect(...)) and another one describing the active neighbors edge case specifically.

Also updated PR summary to be correct. Thanks for calling this out!

mvitousek reviewed (commented):
I like this a lot so far, I think the core algorithm makes a lot of sense and I love the simplifications. Most of my comments are nits, but I want to draw your attention to hte point about union/intersection of neighbor nodes--am I overlooking something? As written it seems like we consider unconditionally accessed nodes to be the *union* of the nodes nodes in a block and the intersection of the nodes accessed in pred/succ blocks. This makes sense to me -- if bbX reads a.b it's safe to read a.b, if bbX's pred and succ read x.b it's safe to read a.b -- but in a couple places I think you describe the algorithm differently.

mvitousek commented:
> However, it may exacerbate [this function-expr hoisting bug](https://github.com/facebook/react/blob/main/compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-invalid-hoisting-functionexpr.ts).

Link is dead

josephsavona reviewed (approved):
A bunch of questions/comments from reading through. Overall it makes sense, definitely very complex. One thing that I think might be a source of incidental complexity is the recursive nature of the approach: propagateScopeDeps() delegates to collectHoistables() which then does collectNodes() and computeSidemaps(), but those inner values from collectHoistables() are used later. 

Seeing it as a sequence of a few high level steps (flatten one level of the function call graph, eg flatten away collectHoistables()) might make this easier to follow.

That said, since this is going to be off by default while you add the remaining pieces, i'm okay with shipping and iterating.

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/visitors.ts:
```diff
@@ -1149,3 +1153,72 @@ export function* eachTerminalOperand(terminal: Terminal): Iterable<Place> {
+export class ScopeBlockTraversal {
+  // Live stack of active scopes
+  #activeScopes: Array<ScopeId> = [];
+  blockInfos: Map<
+    BlockId,
+    | {
+        kind: 'end';
+        scope: ReactiveScope;
+        pruned: boolean;
+      }
+    | {
+        kind: 'begin';
```
A single block can only be either the start of a scope block or the fallthrough. This is because `BuildReactiveScopeTerminals` inserts a scope start block and a scope fallthrough block for every scope. So by construction, a block can be either the start or fallthrough but not both.

```
// before BuildReactiveScopeTerminals
bb0:
[0]
[1]  -┐
[2]   | scope@0
[3]  -┘
...

// after
bb0:
  [0]
  ScopeTerminal block=bb1 fallthr=bb2

bb1: <-- scope begin block
  [1]
  [2]
  Goto bb2

bb2: <-- scope end block
  [3]
  [4]
```

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/DeriveMinimalDependenciesHIR.ts:
```diff
@@ -0,0 +1,267 @@
+
+  addDependency(dep: ReactiveScopePropertyDependency): void {
+    const {path} = dep;
+    let currNode = this.#getOrCreateRoot(dep.identifier, false);
```
We don't know whether the root node (i.e. a named identifier or promoted temporary) is non-null. A node being 'NonNull' means that it's safe to hoist reads from property of that node. (see [recursive deriveMinimalDeps case](https://github.com/facebook/react/blob/999bb6bebcd83570c936fe6f25f0b15de5000140/compiler/packages/babel-plugin-react-compiler/src/HIR/DeriveMinimalDependenciesHIR.ts#L212-L225))

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,520 @@
+  roots: Map<Identifier, RootNode> = new Map();
+
+  #getOrCreateRoot(identifier: Identifier): PropertyLoadNode {
+    // roots can always be accessed unconditionally in JS
```
Updated comment to include this asterisks

mofeiZ commented:
> > However, it may exacerbate [this function-expr hoisting bug](https://github.com/facebook/react/blob/main/compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-invalid-hoisting-functionexpr.ts).
> 
> Link is dead

Updated link (`ts` -> `tsx` 🤦‍♀️)

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,520 @@
+  fn: HIRFunction,
+  nodes: ReadonlyMap<BlockId, BlockInfo>,
+): void {
+  const succ = new Map<BlockId, Set<BlockId>>();
```
Updated to use full names

mofeiZ commented on compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:
```diff
@@ -0,0 +1,520 @@
+      const usedOutside = usedOutsideDeclaringScope.has(
+        lvalue.identifier.declarationId,
+      );
+      if (value.kind === 'PropertyLoad') {
```
Added a TODO for now, will add in a followup PR! Trying to keep the number of semantically meaningful amends small

mofeiZ commented:
Updated per feedback from @mvitousek and @josephsavona. I amended this existing one to keep the blame history clean, but please let me know if you would prefer that I make separate feedback PRs in the future.


Double checked that these changes were no-ops by syncing internally and validating compilation output of 100+k files stayed the same.


Changes:
- rebased on #30949 which forked test fixtures changed by this PR (previously, enablePropagateScopeDepsInHIR was enabled by default for all fixtures)
- refactored to combine `properties: Map<IdentifierId, PropertyLoadNode>` and `temporaries: Map<IdentifierId, Identifier>` into a single `temporaries: Map<IdentifierId, PropertyLoadNode>`. This simplifies a lot of logic as previously we were checking and trying to merge entries from both maps. This should be correct as an lvalue (the identifier id) can only be mapped to a single instruction.
- tweaked recursion logic for `recursivelyDeriveNonNull`.
  - previously, we initially set nonNullObjects = {}, then recursively calculated nonNullObjects = Union(nonNullObjects_own_block, Intersect(nonNullObjects of neighbors))
  - now, initially set nonNullObjects = nonNullObjects_own_block, then recursively calculate nonNullObjects = Union(nonNullObjects, Intersect(nonNullObjects of neighbors))

> One thing that I think might be a source of incidental complexity is the recursive nature of the approach: propagateScopeDeps() delegates to collectHoistables() which then does collectNodes() and computeSidemaps(), but those inner values from collectHoistables() are used later.
- moved `collectSidemap` call out of collectHoistables.
- renamed `collectNodes` -> `collectPropertyLoadsInBlocks`, improved comments, changed recursion to use do-while loop, etc
- added TODO for Destructuring instructions

Should be ready to merge. Thanks again for all the great feedback!

josephsavona commented:
awesome, shipit!

mvitousek reviewed (approved):
Heck yeah
